### PR TITLE
masterfiles-stage.sh --check-only

### PR DESCRIPTION
**Pull request by vpodzime**

### CODE REVIEW

**Review by larsewi — DISMISSED** [review]
🚀

**Review by nickanderson — APPROVED** [review]

**Review by craigcomstock — COMMENTED** [review]
shellcheck would probably say

```
if [ "x$1" = "xtrue" ]; then
     ^---^ SC2268 (style): Avoid x-prefix in comparisons as it no longer serves a purpose.
```

So maybe change that to be tidy.

**Comment by vpodzime** [comment]
> it no longer serves a purpose

How come? I'm not checking `$#` so doing `x$1` is the only way to avoid an empty string there in cases no argument is passed.

**Comment by craigcomstock** [comment]
> How come? I'm not checking $# so doing x$1 is the only way to avoid an empty string there in cases no argument is passed.

`test` aka `[` on all the shells we will use will almost certainly not have this problem of empty string as it did in days past.

https://www.shellcheck.net/wiki/SC2268

What you wrote doesn't hurt, I just like using shellcheck and think it gives generally excellent advice.

**Comment by vpodzime** [comment]
> all the shells we will use will almost certainly not have this problem

Sounds amazing! :grin:

**Review by craigcomstock — DISMISSED** [review]

**Review by mikeweilgart — CHANGES_REQUESTED** [review]
From a code design standpoint this is a major change; I think passing the check_only flag to every single function as an argument is a mistake.  A global variable would be more in keeping, much like the way debug logging is currently handled.  But even better would be to add the necessary new functions and keep them simple.

Overall this seems to make the code more complicated and harder to understand as it adds a second and mutually exclusive path through almost every function, and a separate return point or two.  (Effectively, a whole bunch of "gotos" have been added.)

@tdlee may be interested to review this as well since his use of this code is more regular than mine.

File: contrib/masterfiles-stage/common.sh [hunk 1]
```diff
@@ -287,9 +295,31 @@ git_cfbs_deploy_refspec() {
...
+  # $2 -- checked out work-tree
+  # $3 -- refspec
+
+  # Get the hash of the refspec in the mirror and compare it to the checked out
+  # work-tree (see git_deploy_refspec() for details).
+  # NOTE: 'format:%H' usually results in just one line, but in case of a tag,
+  #       the output contains info about the tag followed by the hash on the
+  #       last line
+  mirror_rev="$(git --git-dir="$1" show --pretty=format:%H -s "$3" | tail -n1)"
+  work_tree_rev="$(cat "$2/.git/HEAD")"
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply contrib/masterfiles-stage/common.sh:309]: yup, it's used by `cf-promises -T` so it gets copied over.

**Comment by vpodzime** [comment]
> From a code design standpoint this is a major change; I think passing the check_only flag to every single function as an argument is a mistake. A global variable would be more in keeping, much like the way debug logging is currently handled. But even better would be to add the necessary new functions and keep them simple.

I'd argue that using global variables is a mistake. :wink: :slightly_smiling_face: But since they are already used extensively, adding one more would not be a problem. @nickanderson @olehermanse opinions?

Regarding new simple functions -- yes, that would be nice for sure, but since the logic of what is being checked out and where is buried deep in the functions we have, I'd have to reimplement the same logic. That's why it made more sense to me to keep the logic in one place and then just do a different thing when only checking for updates.

**Comment by vpodzime** [comment]
> @tdlee may be interested to review this as well since his use of this code is more regular than mine.

yup! The more :eyes: the better!

File: contrib/masterfiles-stage/common.sh [hunk 2]
```diff
@@ -287,9 +295,31 @@ git_cfbs_deploy_refspec() {
...
+  # $1 -- git repo mirror
+  # $2 -- checked out work-tree
+  # $3 -- refspec
+
+  # Get the hash of the refspec in the mirror and compare it to the checked out
+  # work-tree (see git_deploy_refspec() for details).
+  # NOTE: 'format:%H' usually results in just one line, but in case of a tag,
+  #       the output contains info about the tag followed by the hash on the
+  #       last line
+  mirror_rev="$(git --git-dir="$1" show --pretty=format:%H -s "$3" | tail -n1)"
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply contrib/masterfiles-stage/common.sh:308]: `^0` works nicely. Thanks!

**Comment by olehermanse** [comment]
> > How come? I'm not checking $# so doing x$1 is the only way to avoid an empty string there in cases no argument is passed.
> 
> `test` aka `[` on all the shells we will use will almost certainly not have this problem of empty string as it did in days past.
> 
> https://www.shellcheck.net/wiki/SC2268
> 
> What you wrote doesn't hurt, I just like using shellcheck and think it gives generally excellent advice.

@craigcomstock @vpodzime FYI, I followed that link, and then the link to a related resource, which says;

> I was unfortunately unable to peer into the closely guarded world of the likes of HP-UX and AIX, so dinosaur herders beware.

https://www.vidarholen.net/contents/blog/?p=1035

So....

**Comment by nickanderson** [comment]
> I'd argue that using global variables is a mistake. 😉 🙂 But since they are already used extensively, adding one more would not be a problem. @nickanderson @olehermanse opinions?

I have no strong opinion here.  Your change while perhaps not ideal is still small. We already have other duplication. I would say a great refactor would be a separate issue.

**Review by craigcomstock — APPROVED** [review]

**Review by olehermanse — APPROVED** [review]

**Review by larsewi — APPROVED** [review]

**Pull request merged**